### PR TITLE
Issue #3090450 by mdeny, agami4: As a LU I want to reach the comment I wanted to interact with when I had to join a group

### PR DESCRIPTION
**Pull request by denysyukm**
## Problem
As a LU I want to reach the comment I wanted to interact with when I had to join a group

## Solution
Implement new flow joining the groups from comments section

## Issue tracker
https://www.drupal.org/project/social/issues/3090450
https://getopensocial.atlassian.net/browse/YANG-1622
https://getopensocial.atlassian.net/browse/YANG-1719

## How to test
Enable group_core_comments
- [ ] as a LU go to the group you aren't joined
- [ ] go to the content of the group
- [ ] below comments section click `here` in the text: "You are not allowed to comment on content in a group you are not member of. You can join the group here."
- [ ] popup with group info should appear
- [ ] on the popup click "Join group"
- [ ] you should be joined
- [ ] you should be on the group content page where you wanted to interact with
- [ ] you should have the possibility to post the comments

## Release notes
Implement new flow to join the groups when users want to reach the comment they wanted to interact with when they had to join a group

### CODE REVIEW

File: modules/custom/group_core_comments/assets/css/group_core_comments.scss [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      height: 0;
+      width: 0;
+      position: absolute;
+      left: 50%;
+      pointer-events: none;
+      transform: translateX(-50%);
+    }
+
+    &:before {
+      border-top-color: rgba(0,0,0,0.12);
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/css/group_core_comments.scss]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+(function ($) {
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+      $(document).click(function(event) {
+        if ($(event.target).closest('.forbidden-post-comments-wrapper .description .btn-action__group').length) return;
+        if ($(event.target).closest('.forbidden-post-comments-wrapper .popup-info').length) return;
+        link.removeClass('open');
+        event.stopPropagation();
+      });
+    }
+  };
+})(jQuery);
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        popup.css('top', (-popupHResize - 5));
+      });
+
+      link.on('click', function (e) {
+        e.preventDefault();
+        $(this).toggleClass('open');
+      });
+
+      $(document).click(function(event) {
+        if ($(event.target).closest('.forbidden-post-comments-wrapper .description .btn-action__group').length) return;
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        if ($(event.target).closest('.forbidden-post-comments-wrapper .popup-info').length) return;
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      var link = forbiddenPost.find('.description .btn-action__group');
+
+      popup.css('top', (-popupH - 5));
+
+      $(window).on('resize', function () {
+        var popupHResize = popup.outerHeight();
+        popup.css('top', (-popupHResize - 5));
+      });
+
+      link.on('click', function (e) {
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        e.preventDefault();
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        $(this).toggleClass('open');
+      });
+
+      $(document).click(function(event) {
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Fixed

File: modules/custom/group_core_comments/assets/js/group_core_comments.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    attach: function (context) {
+      var forbiddenPost = $('.forbidden-post-comments-wrapper', context);
+      var popup = forbiddenPost.find('.popup-info', context);
+      var popupH = popup.outerHeight();
+      var link = forbiddenPost.find('.description .btn-action__group');
+
+      popup.css('top', (-popupH - 5));
+
+      $(window).on('resize', function () {
+        var popupHResize = popup.outerHeight();
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/assets/js/group_core_comments.js]: Const does not work without bable

File: modules/custom/group_core_comments/group_core_comments.libraries.yml [hunk 10]
```diff
@@ -0,0 +1,6 @@
+group_core_comments:
+  css:
+    theme:
+      assets/css/group_core_comments.css: {}
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/group_core_comments.libraries.yml]: I had small discussed with Alexandr about name folders. And nesting of folders must be `assets/css or js`

File: modules/custom/group_core_comments/group_core_comments.libraries.yml [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+  js:
+    assets/js/group_core_comments.js: {}
```
[the remark this answers is not in the document]
Comment (agami4) [reply modules/custom/group_core_comments/group_core_comments.libraries.yml]: I had small discussed with Alexandr about name folders. And nesting of folders must be `assets/css or js`

File: modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+ * Provides group core comments controllers.
+ */
+class GroupCoreCommentsController extends ControllerBase {
+
+  /**
+   * Entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
```
[the remark this answers is not in the document]
Comment (denysyukm) [reply modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php]: Deleted

File: modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+  /**
+   * Current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $currentUser;
```
[the remark this answers is not in the document]
Comment (denysyukm) [reply modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php]: Deleted

File: modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function quickJoinGroup(GroupInterface $group) {
+    /** @var \Drupal\group\Plugin\GroupContentEnablerInterface $plugin */
+    $plugin = $group->getGroupType()->getContentPlugin('group_membership');
+
+    $group_content = $this->entityTypeManager->getStorage('group_content')->create([
```
[the remark this answers is not in the document]
Comment (denysyukm) [reply modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php]: Fixed

File: modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+      'type' => $plugin->getContentTypeConfigId(),
+      'gid' => $group->id(),
+      'entity_id' => $this->currentUser->id(),
```
[the remark this answers is not in the document]
Comment (denysyukm) [reply modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php]: DONE

File: modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    $this->entityTypeManager = $entity_type_manager;
+    $this->currentUser = $current_user;
+    $this->requestService = $request_stack;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
```
[the remark this answers is not in the document]
Comment (denysyukm) [reply modules/custom/group_core_comments/src/Controller/GroupCoreCommentsController.php]: Added

**Comment by ronaldtebrake** [comment]
Hi guys, functional wise I'm happy, could you rebase this so travis works?
Also if you have time (but I don't consider this a blocker // can accept a follow up)
Screen Shot 2020-01-22 at 12 59 42
Can you have a look at the design when there are no comments?

**Comment by ronaldtebrake** [comment]
Fixed it myself so it can be part of 8.0 after travis is running green, also no impact for the feature removal. 
Might be a good idea to check if we want to enable this module by default, I think it has a lot of added benefit and can easily be done so in a update hook of social_group

**Pull request merged**